kcbq-api/src/main/java/com/wepay/kafka/connect/bigquery/api/SchemaRetriever.java

```diff
@@ -21,17 +23,19 @@
   /**
    * Retrieve the most current schema for the given topic.
    * @param table The table that will be created.
-   * @param topic The topic to retrieve a schema for.
+   * @param topicAndRecordName The topic and an optional record name to retrieve a schema for.
    * @param schemaType The type of kafka schema, either "value" or "key".
    * @return The Schema for the given table.
    */
-  public Schema retrieveSchema(TableId table, String topic, KafkaSchemaRecordType schemaType);
+  public Schema retrieveSchema(TableId table, TopicAndRecordName topicAndRecordName, KafkaSchemaRecordType schemaType);
+
+  Map<TopicAndRecordName, Schema> retrieveSchemas(List<String> topics, Map<Pattern, String> recordAliases);
```

Comment:
This method is aimed to pre-fetch all the current schemas from registry in order to create all the necessary tables. 

If you take a look on the implementation you'll see that I'm fetching all subjects and then do some pattern matching. I do this because number of types of events (record types) might be humongous and increase in time (usually new feature introduces a couple of events). Maintaining a list of all supported record types for a topic in the connector's configuration would not only be cumbersome but also would require restarts and reconfiguration whenever you introduce new record type (which becomes more problematic when using some hosted kafka solutions).

That's why I decided to rely on regular expressions which let me declare one or a couple of patterns (instead of dozens or hundreds) to rule them all. Also since creating missing tables on the fly is supported, there is no need for restarts.

Unfortunately we cannot query schema reqistry using those patterns and we do not know exact record names in advance.
The only solution is to fetch them all and filter, using regular expressions defined in the config.

Alternatively we could just forget about creating tables at startup time and create them on the fly, when a new record type appears. This is something @bingqinzhou introduced recently in #233. If all tables would be created on demand and there won't be need for pre-fetching subjects, then we could get rid of this method and also make several classes (including TopicAndRecordName) much simpler.

WDYT?

Reply:
> Alternatively we could just forget about creating tables at startup time and create them on the fly, when a new record type appears. This is something @bingqinzhou introduced recently in #233. If all tables would be created on demand and there won't be need for pre-fetching subjects, then we could get rid of this method and also make several classes (including TopicAndRecordName) much simpler.

Aha, thanks for the clarification. Creating table on the fly is really what we want to have (and we just added recently). So, yes, I prefer that over creating tables at startup. Sounds like it will also make your implementation easier. (which is great!)